devicetypes/shinasys/sihas-multipurpose-sensor.src/sihas-multipurpose-sensor.groovy

```diff
@@ -0,0 +1,299 @@
+                            [value: 59, color: "#90d2a7"],
+                            [value: 74, color: "#44b621"],
+                            [value: 84, color: "#f1d801"],
+                            [value: 95, color: "#d04e00"],
+                            [value: 96, color: "#bc2323"]
+                    ]
+            )
+        }
+        valueTile("humidity", "device.humidity", decoration: "flat", inactiveLabel: false, width: 2, height: 2) {
+            state "humidity", label: '${currentValue}% humidity', unit: ""
+        }
+        
+        valueTile("illuminance", "device.illuminance", width: 2, height: 2) {
+            state "luminosity", label:'${currentValue} ${unit}', unit:"lux"
+        }
+        
+        valueTile("battery", "device.battery", decoration: "flat", inactiveLabel: false, width: 2, height: 2) {
+            state "battery", label: '${currentValue}% battery', unit: ""
+        }
+        standardTile("refresh", "device.refresh", inactiveLabel: false, decoration: "flat", width: 2, height: 2) {
+            state "default", action: "refresh.refresh", icon: "st.secondary.refresh"
+        }
+    
+        main(["motion", "temperature"])
+        details(["motion", "temperature", "humidity", "illuminance", "battery", "refresh"])
+    }
+}
+
+private getILLUMINANCE_MEASUREMENT_CLUSTER() { 0x0400 }
+private getOCCUPANCY_SENSING_CLUSTER() { 0x0406 }
+private getPOWER_CONFIGURATION_BATTERY_VOLTAGE_ATTRIBUTE() { 0x0020 }
+private getTEMPERATURE_MEASUREMENT_MEASURED_VALUE_ATTRIBUTE() { 0x0000 }
+private getRALATIVE_HUMIDITY_MEASUREMENT_MEASURED_VALUE_ATTRIBUTE() { 0x0000 }
+private getILLUMINANCE_MEASUREMENT_MEASURED_VALUE_ATTRIBUTE() { 0x0000 }
+private getOCCUPANCY_SENSING_OCCUPANCY_ATTRIBUTE() { 0x0000 }
+
+private List<Map> collectAttributes(Map descMap) {
+    List<Map> descMaps = new ArrayList<Map>()
+
+    descMaps.add(descMap)
+
+    if (descMap.additionalAttrs) {
+        descMaps.addAll(descMap.additionalAttrs)
+    }
+
+    return  descMaps
+}
+
+def parse(String description) {
+    Map map = zigbee.getEvent(description)
+    if (!map) {
+        if (description?.startsWith('zone status')) {
+            map = parseIasMessage(description)
+        } else if (description?.startsWith('read attr')) {
+            Map descMap = zigbee.parseDescriptionAsMap(description)
+
+            if (descMap?.clusterInt == zigbee.POWER_CONFIGURATION_CLUSTER && descMap.commandInt != 0x07 && descMap.value) {
+                List<Map> descMaps = collectAttributes(descMap)
+                def battMap = descMaps.find { it.attrInt == POWER_CONFIGURATION_BATTERY_VOLTAGE_ATTRIBUTE }
+
+                if (battMap) {
+                    map = getBatteryResult(Integer.parseInt(battMap.value, 16))
+                }
+            } else if (descMap?.clusterInt == zigbee.IAS_ZONE_CLUSTER && descMap.attrInt == zigbee.ATTRIBUTE_IAS_ZONE_STATUS && descMap.commandInt != 0x07) {
```

Comment:
Again, this condition and the one from line 123 seem redundant to each other. My guess is, if with current code DTH handles devices correctly, you can remove second condition (the one in line 123), cause, additionally, it seems that you're converting `descMap.value` to integer using 10 as a number base, which is wrong.

Reply:
I made it by referring to smartsense-multi-sensor.groovy.
Anyway, it seems to be the wrong code.
I committed again.
I have an additional question.
I applied the vid to each device in fingerprint code, did I apply it properly?
 * model "USM-300Z"  : Temperature,Humidity,Occupancy,Illuminance sensor, Battery
 * model "OSM-300Z"  : Occupancy sensor, Battery
 * model "TSM-300Z"  : Temperature,Humidity sensor, Battery

Reply:
Yeah, those definitions seems to be correct.